src/controller/src/lib.rs

```diff
@@ -364,10 +364,25 @@ where
                             memory_limit: size_config.memory_limit,
                             scale: size_config.scale,
                             labels: hashmap! {
+                                "replica-id".into() => replica_id.to_string(),
                                 "cluster-id".into() => instance_id.to_string(),
                                 "type".into() => "cluster".into(),
                             },
                             availability_zone,
+                            anti_affinity: Some(vec![
+                                LabelSelector {
+                                    label_name: "cluster-id".to_string(),
+                                    logic: LabelSelectionLogic::Eq {
+                                        value: instance_id.to_string(),
+                                    },
+                                },
+                                LabelSelector {
+                                    label_name: "replica-id".into(),
+                                    logic: LabelSelectionLogic::NotEq {
+                                        value: replica_id.to_string(),
+                                    },
+                                },
+                            ]),
```

Comment:
Just to check my understanding: These selectors mean "This pod should not run on any node that already has a pod from the same cluster running" and "This pod should not run on any node that doesn't already have a pod from the same replica running". So we try to schedule pods of a replica together, and pods of different replicas in the same cluster apart.

Do you know how these rules interact? They are somewhat in conflict with each other. It would be nice if we could somehow make sure the "schedule pods of the same replica together" one has more weight, to prevent k8s from needlessly pulling our replicas apart by following the first rule.

Reply:
Thanks! Yeah, my misunderstanding was thinking that the selectors are ORed together. Brennan educated me yesterday, but I forgot to mark this discussion as resolved.

> Separately, it might make sense for multi-process replicas to have an affinity preference for running them on the same node to reduce the number of network / host failures that can happen

This makes sense. Although I think currently our strategy is to grow to the largest possible single-pod replica before resorting to multiple pods. Which means if we have a replica consisting of multiple pods, these should never fit on a single node. I might be mixing something up here though.